herddb-core/src/main/java/herddb/core/DBManager.java

```diff
@@ -1186,18 +1188,17 @@ private boolean manageTableSpaces() {
             try {
                 for (TableSpaceManager tableSpaceManager : followingActiveTableSpaces) {
                     String tableSpaceUuid = tableSpaceManager.getTableSpaceUUID();
-                    TableSpace tableSpaceInfo = actualTableSpaceMetadata.get(tableSpaceUuid);
+                    TableSpace tableSpaceInfo = currentTableSpaceMetadata.get(tableSpaceUuid);
 
                     if (tableSpaceInfo != null
+                            && !tableSpaceInfo.leaderId.equals(nodeId)
                             && tableSpaceInfo.maxLeaderInactivityTime > 0
                             && !tableSpaceManager.isFailed()) {
                         List<TableSpaceReplicaState> allReplicas =
                                 metadataStorageManager.getTableSpaceReplicaState(tableSpaceUuid);
                         TableSpaceReplicaState leaderState = allReplicas
                                 .stream()
-                                .filter(t -> t.mode == TableSpaceReplicaState.MODE_LEADER
-                                        && !t.nodeId.equals(nodeId)
-                                )
+                                .filter(t -> t.nodeId.equals(tableSpaceInfo.leaderId))
```

Comment:
This is the core of the change.
In case of multiple servers trying to become leader the prev check could return values from the dead leader resulting in lots of leadership changes.